examples/sunlinsol/ginkgo/test_sunlinsol_ginkgo.cpp

```diff
@@ -0,0 +1,585 @@
+   * Create system matrix *
+   * -------------------- */
+
+  /* Wrap ginkgo matrix for SUNDIALS,
+     Matrix is overloaded to a SUNMatrix. */
+
+  std::unique_ptr<sundials::ConvertibleTo<SUNMatrix>> A;
+
+  auto matrix_dim{gko::dim<2>(matrows, matcols)};
+
+  if (matrix_type == "csr") {
+    using GkoMatrixType = gko::matrix::Csr<sunrealtype, sunindextype>;
+    auto matrix_nnz{3 * matrows - 2};
+    auto gko_matrix = gko::share(GkoMatrixType::create(gko_exec, matrix_dim, matrix_nnz));
+
+    if (matcond) {
+      auto gko_matdata{gko::matrix_data<sunrealtype, sunindextype>::cond(matrows,
+                                                                         gko::remove_complex<sunrealtype>{matcond},
+                                                                         distribution_real, engine)};
+      gko_matdata.remove_zeros();
+      gko_matrix->read(gko_matdata);
+    }
+    else {
+      fill_matrix(gko::lend(gko_matrix));
+    }
+    A = std::make_unique<sundials::ginkgo::Matrix<GkoMatrixType>>(std::move(gko_matrix), sunctx);
+  }
+  else if (matrix_type == "dense") {
+    using GkoMatrixType = gko::matrix::Dense<sunrealtype>;
+    auto gko_matrix     = gko::share(GkoMatrixType::create(gko_exec, matrix_dim));
+    if (matcond) {
+      auto gko_matdata{gko::matrix_data<sunrealtype, sunindextype>::cond(matrows,
+                                                                         gko::remove_complex<sunrealtype>{matcond},
+                                                                         distribution_real, engine)};
+      gko_matdata.remove_zeros();
+      gko_matrix->read(gko_matdata);
+    }
+    else {
+      gko_matrix->fill(0.0);
+      fill_matrix(gko::lend(gko_matrix));
+    }
+    A = std::make_unique<sundials::ginkgo::Matrix<GkoMatrixType>>(std::move(gko_matrix), sunctx);
+  }
+
+  /* Create right-hand side vector for linear solve */
+  fails += SUNMatMatvecSetup(A->Convert());
+  fails += SUNMatMatvec(A->Convert(), x, b);
+  if (fails) {
+    std::cerr << "FAIL: SUNLinSol SUNMatMatvec failure\n";
+    N_VDestroy(x);
+    N_VDestroy(b);
+    return 1;
+  }
+
+  /* -------------------- *
+   * Create linear solver *
+   * -------------------- */
+
+  /* Use default stopping criteria */
+  auto crit{sundials::ginkgo::DefaultStop::build().with_max_iters(max_iters).on(gko_exec)};
+
+  /* Use a Jacobi preconditioner */
+  auto precon{gko::preconditioner::Jacobi<sunrealtype, sunindextype>::build().on(gko_exec)};
+
+  /* Wrap ginkgo matrix for SUNDIALS,
+     Matrix is overloaded to a SUNLinearSolver. */
+  std::unique_ptr<sundials::ConvertibleTo<SUNLinearSolver>> LS;
+
+  if (method == "bicg") {
+    using GkoSolverType = gko::solver::Bicg<sunrealtype>;
+    auto gko_solver_factory{
+        GkoSolverType::build().with_criteria(std::move(crit)).with_preconditioner(std::move(precon)).on(gko_exec)};
+    if (matrix_type == "csr") {
+      using GkoMatrixType = gko::matrix::Csr<sunrealtype, sunindextype>;
```

Comment:
Unfortunately, your suggestion won't work because `using` statements are scoped to the block they are created in. So `GkoMatrixType` would not be defined in the second or third block.